### DISCUSSION TITLE
Example of using Client Side Blazor with Identity and CSLA

### DISCUSSION THREAD

**Discussion by mtavares628** [MarimerLLC/csla#1956] [Questions]
I'm wondering if someone could provide an example of how to connect up the client side CslaUserService with the ClaimsPrincipal using .NET Core Identity.  I've seen the client side authentication example in the Blazor book as well as the ProjectTracker application where the Login page is on the client, it uses the dataportal to call the data access layer to get the user and role info, and then takes that info and generates the ClaimsPrincipal on the client side and assign it to the UserService.  

I was also looking at what the Client Side Blazor template does when you add authentication when creating the solution.  It generates the login page on the app server (similar to server side Blazor) and will use the SignInManager of the Identity framework to perform the login.  

What I can't wrap my head around is how I could utilize the Identity Framework on the server to send that ClaimsPrincipal back to the client if I were to utilize the template's server side login functionality.  I'd like to take advantage of some of the built in functionality of Identity (like Lockout and 2FA) rather than rolling my own by building upon the ProjectTracker example.  

Any help would be greatly appreciated. Thanks.

**Comment by rockfordlhotka** [comment]
I do have a server-side Blazor project that uses the aspnet user database concept.

I don't have a client-side Blazor project that does this.

I think the important thing is to understand the purpose of the AuthenticationStateProvider and user service.

Blazor doesn't rely on the thread's CurrentPrincipal or HttpContext to maintain the user's identity. Instead, they externalized that concept to your `AuthenticationStateProvider`. By convention, that provider typically relies on some sort of user service.

The `CslaAuthenticationStateProvider` and `CslaUserService` are very simple implementations of these concepts to support the simplest scenario, relying on `Csla.ApplicationContext.User` to maintain the current user identity in memory.

At the end of the day, CSLA itself relies on `Csla.ApplicationContext.User` having the current user identity. If the user principal is _stored_ elsewhere (like in some other user service), then you need to create a custom CSLA application context manager that gets/sets the principal from that user service instead of using the default CSLA behavior.

**Reply by mtavares628** [reply]
Thanks for the info Rocky.  I am sort of working on a hybrid of your client example where instead of hitting the DB in the dataportal to grab the Principle/Identity info, I'm using SignInManager and UserManager.  This way it handles the password hashing and everything else and then just passes down the ClaimsPrincipal.  However, I've run into an issue where the ClaimsPrincipal isn't serializing.  I saw in some other posts that MobileFormatter should be able to serialize the ClaimsPrinciple, but when I run I get the following:

```
System.NotSupportedException: Serialization of type object is not supported. (System.Security.Claims.ClaimsPrincipal) blazor.webassembly.js:1:38061
   at Csla.Serialization.Mobile.CslaBinaryWriter.Write(Object target, BinaryWriter writer)
blazor.webassembly.js:1:38061
   at Csla.Serialization.Mobile.CslaBinaryWriter.Write(Stream serializationStream, List`1 objectData)
blazor.webassembly.js:1:38061
   at Csla.Serialization.Mobile.MobileFormatter.Serialize(Stream serializationStream, Object graph)
blazor.webassembly.js:1:38061
   at Csla.Serialization.Mobile.MobileFormatter.Serialize(Object obj)
blazor.webassembly.js:1:38061
   at Csla.Server.Hosts.HttpPortal.Fetch(CriteriaRequest request)
blazor.webassembly.js:1:38061
   at Csla.Server.Hosts.HttpPortalController.CallPortal(String operation, Object request)
blazor.webassembly.js:1:38061
   at Csla.Server.Hosts.HttpPortalController.InvokeTextPortal(String operation, Stream requestStream, Stream responseStream)
```

Should the MobileFormatter work in this case?  As a workaround for now, I've just done what you have and passed down the user and role info, and created the ClaimsPrincipal on the client. But I'm just curious on whether I should be able to send the ClaimsPrincipal without issue.  Do I have to add anything to the config on the client config?

My Main currently looks like:
`public static async Task Main(string[] args)
        {
            var builder = WebAssemblyHostBuilder.CreateDefault(args);
            builder.RootComponents.Add<App>("app");

            builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });

            builder.Services.AddAuthorizationCore();
            builder.Services.AddOptions();
            builder.Services.AddSingleton
              <AuthenticationStateProvider, CslaAuthenticationStateProvider>();
            builder.Services.AddSingleton<CslaUserService>();
            App.IsServerSide = false;
            builder.UseCsla(c => c
                .DataPortal()
                .DefaultProxy(typeof(Csla.DataPortalClient.HttpProxy),
                              "https://localhost:44326/api/dataportaltext/")
                );
            
            await builder.Build().RunAsync();
        }
    }`

and my Startup.cs on the App Server has:
`app.UseCsla(config => config.Configure()
                            .Serialization()
                            .SerializationFormatter("MobileFormatter")`

**Reply by rockfordlhotka** [reply]
The MobileFormatter (recently) includes support to identify that a `ClaimsPrincipal` is in the object graph, and to serialize it. I'm pretty confident that this works, as it is used in the [CslaAuthentication example](https://github.com/MarimerLLC/BlazorBook/tree/master/BlazorCslaAuthentication) in my Blazor book. The live code in that repo has been upgraded to .NET 5 and a 5.4.0 prerelease of CSLA (I still need to update it to the 5.4.0 release).

**Update:** OK, so looking at the sample, it doesn't actually have any functionality to try and transfer the principal through the data portal from client to server.

[ProjectTracker](https://github.com/MarimerLLC/csla/tree/master/Samples/ProjectTracker) does though, because it is a complete implementation, and the Blazor wasm client does rely on `ClaimsPrincipal`.

**Reply by mtavares628** [reply]
And I'm actually following the ProjectTracker model as a workaround.  What I was trying to do was go in the opposite direction.  Create the ClaimsPrincipal in the Dataportal on the Server side and send it back down as a property on the UserDto. So my UserDal looks like this:

```
public class UserDal : IUserDal
    {
        private readonly SignInManager<DadUser> _signInManager;

        public UserDal(SignInManager<DadUser> signInManager)
        {
            _signInManager = signInManager;
        }

        public UserDto Fetch(string username, string password, int advocacyGroupId)
        {
            // Combine to inject both the username and the advocacy group into the username field
            string usernameAndGroup = string.Format("{0}|{1}", username, advocacyGroupId);
            var result = _signInManager.PasswordSignInAsync(usernameAndGroup, password, false, lockoutOnFailure: false).Result;

            if (result.Succeeded)
            {
                var user = _signInManager.UserManager.FindByNameAsync(usernameAndGroup).Result;
                if (user != null)
                {
                    var roles = _signInManager.UserManager.GetRolesAsync(user).Result;

                    var principal = _signInManager.CreateUserPrincipalAsync(user).Result;
                    
                    return new UserDto
                    {
                        AdvocacyGroupID = advocacyGroupId,
                        Username = username,
                        EmployeeID = user.Id,
                        Roles = roles,
                        Principal = principal
                        
                    };
                }
                else return null;
            }
            else return null;
        }

        public UserDto Fetch(string username)
        {
            throw new NotImplementedException();
        }
    }
```

And then the fetch in my CredentialValidator looks like:
```
[Fetch]
        private void Fetch(Credentials credentials, [Inject] IUserDal dal)
        {
            var data = dal.Fetch(credentials.Username, credentials.Password, credentials.AdvocacyGroupID);
            if (data != null)
            {
                AuthenticationType = "Identity.Application";
                UserPrincipal = data.Principal;
                Name = data.Username;
                AdvocacyGroupID = data.AdvocacyGroupID;
                EmployeeID = data.EmployeeID;
                Roles = new MobileList<string>(data.Roles);
            }
        }
```

And the Login.razor LoginUser method just sets the CslaUserService user to the returned principal instead of creating it on the client side like the Project Tracker does:
```
 private async void LoginUser()
    {
        vm.Model.AdvocacyGroupID = 1;
      var validator =
        await Csla.DataPortal.FetchAsync<CredentialValidator>(vm.Model);
      var principal = validator.UserPrincipal();

      if (principal.Identity.IsAuthenticated)
      {
        userService.CurrentUser = principal;
        NavigationManager.NavigateTo("/");
      }
      else
      {
        ErrorText = "Invalid credentials";
        StateHasChanged();
      }
    }
```

Am I just totally wrong in thinking that the DataPortal would serialize the ClaimsPricipal going in the other direction?

**Reply by rockfordlhotka** [reply]
I see what you are saying.

It _should_ be the case that MobileFormatter can serialize `ClaimsPrincipal` - though I know there's [a bug with netfx](https://github.com/MarimerLLC/csla/issues/1877) (so if your server is .NET 4.x it won't work). It _should_ work if your client is Blazor and server is netcore or net5.

Of course there could be a bug 😳

**Reply by mtavares628** [reply]
My server is currently netcore3.1.  I'll try updating everything to net5 and see if it's still an issue.  I'll let you know how I make out.  Thanks.

**Reply by rockfordlhotka** [reply]
Hmm. So I added this test

```c#
  [TestClass]
  public class ClaimsPrincipalTests
  {
    [TestMethod]
    public void CloneClaimsPrincipal()
    {
      Configuration.ConfigurationManager.AppSettings["CslaSerializationFormatter"] = "MobileFormatter";
      var i = new ClaimsIdentity();
      i.AddClaim(new Claim("name", "Franklin"));
      var p = new ClaimsPrincipal(i);
      var p1 = (ClaimsPrincipal)Core.ObjectCloner.Clone(p);
      Assert.AreNotSame(p, p1, "Should be different instances");
      Assert.AreEqual(p.Claims.Count(), p1.Claims.Count(), "Should have same number of claims");
      var c = p1.Claims.Where(r => r.Type == "name").First();
      Assert.AreEqual("Franklin", c.Value, "Claim value should match");
    }
  }
```

And it passes, thus establishing that `MobileFormatter` does correctly serialize a `ClaimsPrincipal` as it should.
